File: app/src/main/java/com/mikifus/padland/NewPadActivity.java
package com.mikifus.padland;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.Menu;
import android.view.View;
import android.webkit.URLUtil;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;

import com.mikifus.padland.Models.Server;
import com.mikifus.padland.Models.ServerModel;
import com.mikifus.padland.Utils.PadUrl;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;

/**
 * Allows the user to create a new pad, choosing a name and the host.
 *
 * @author mikifus
 */
public class NewPadActivity extends PadLandActivity {

    protected String[] server_list;
    protected String[] server_url_list;
    protected String[] server_url_prefixed_list;
    protected ServerModel serverModel;

    /**
     * onCreate override
     * @param savedInstanceState
     */
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_newpad);

        server_list = getSpinnerValueList();
        serverModel = new ServerModel(this);
        server_url_list = serverModel.getServerUrlList(NewPadActivity.this);
        server_url_prefixed_list = serverModel.getServerUrlPrefixList(this);

        _setViewEvents();
        _setSpinnerValues();
        _setSpinnerDefaultValue();
    }

    private void _setViewEvents() {
        final EditText nameInput = findViewById( R.id.editText );
        nameInput.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void afterTextChanged(Editable s) {
                String name = s.toString();
                if(URLUtil.isValidUrl(name)) {
                    Uri uri = Uri.parse(name);
                    String padServer = uri.getScheme() + "://" + uri.getHost();
                    String padName = uri.getLastPathSegment();

                    int c = 0;
                    for (String serverUrl : server_url_list) {
                        if(serverUrl.equals(padServer)) {
                            nameInput.setText(padName);

                            Spinner spinner = findViewById( R.id.spinner );https://etherpad.wikimedia.org/p/anabelle
                            spinner.setSelection(c);

                            Toast.makeText(NewPadActivity.this, getString(R.string.newpad_url_name_success), Toast.LENGTH_LONG).show();
                            return;
                        }
                        c++;
                    }
                    Toast.makeText(NewPadActivity.this, getString(R.string.newpad_url_name_warning), Toast.LENGTH_LONG).show();
                }
            }
        });
    }

    /**
     * Returns a string with the server names.
     * Includes custom servers.
     * @return String[]
     */
    private String[] getSpinnerValueList() {
        String[] server_list;
        // Load the custom servers
        ServerModel serverModel = new ServerModel(this);
        ArrayList<Server> custom_servers = serverModel.getEnabledServerList();
        ArrayList<String> server_names = new ArrayList<>();
        for(Server server : custom_servers) {
            server_names.add(server.getName());
        }

        // Server list to provide a fallback value
//        server_list.getResources().getStringArray( R.array.etherpad_servers_name );
        Collection<String> collection = new ArrayList<>();
        collection.addAll(server_names);
        collection.addAll(Arrays.asList(getResources().getStringArray( R.array.etherpad_servers_name )));

        server_list = collection.toArray(new String[collection.size()]);

        return server_list;
    }

    /**
     * Loads the values on the Spinner which is by deafult empty.
     */
    private void _setSpinnerValues() {
        Spinner spinner = (Spinner) findViewById( R.id.spinner );
        if (spinner != null) {
            //selected item will look like a spinner set from XML
            ArrayAdapter<String> spinnerArrayAdapter = new ArrayAdapter<>(this, android.R.layout.simple_spinner_dropdown_item, server_list);
            spinner.setAdapter(spinnerArrayAdapter);
        }
    }

    /**
     * Gets the default value from user settings and selects it in the
     * spinner. This value can be changed in the settings activity.
     */
    private void _setSpinnerDefaultValue(){
        String default_server = this._getDefaultSpinnerValue();

        // We get position and set it as default
        Spinner spinner = findViewById( R.id.spinner );
        ArrayAdapter adapter = (ArrayAdapter) spinner.getAdapter();
        spinner.setSelection( adapter.getPosition( default_server ) );
    }

    /**
     * Returns a string with the default server name. Being a key in the server
     * array
     * @return
     */
    private String _getDefaultSpinnerValue(){
        // Getting user preferences
        Context context = getApplicationContext();
        SharedPreferences userDetails = context.getSharedPreferences( getPackageName() + "_preferences", context.MODE_PRIVATE );
        String default_server = userDetails.getString( "padland_default_server", server_list[0] );
        return default_server;
    }

    /**
     * Creates the menu
     * @param menu
     * @return
     */
    public boolean onCreateOptionsMenu( Menu menu ) {
        return super.onCreateOptionsMenu(menu, R.menu.new_pad);
    }

    /**
     * Form submit
     * @param w
     */
    public void onCreateButtonClick( View w ){
        String padName = this.getPadNameFromInput((TextView) findViewById(R.id.editText));
        String padLocalName = this.getPadNameFromInput((TextView) findViewById(R.id.editTextLocalName));
        String padPrefix = this.getPadPrefixFromSpinner((Spinner) findViewById(R.id.spinner));
        String padServer = this.getPadServerFromSpinner( (Spinner) findViewById(R.id.spinner) );

        Log.d("CREATENEW", padName);
        if ( padName.isEmpty() ) {
            Toast.makeText(this, (getString(R.string.newpad_noname_warning)), Toast.LENGTH_LONG).show();
            return;
        }

        PadUrl padUrl = new PadUrl.Builder()
                .padName(padName)
                .padServer(padServer)
                .padPrefix(padPrefix)
                .build();

        if( !URLUtil.isValidUrl(padUrl.getString()) )
        {
            Toast.makeText(this, getString(R.string.new_pad_name_invalid), Toast.LENGTH_LONG).show();
            return;
        }

        Intent padViewIntent =
                new Intent( NewPadActivity.this, PadViewActivity.class );
        padViewIntent.putExtra( "padName", padUrl.getPadName() );
        padViewIntent.putExtra( "padLocalName", padLocalName );
        padViewIntent.putExtra( "padServer", padUrl.getPadServer() );
        padViewIntent.putExtra( "padUrl", padUrl.getString() );

        startActivity(padViewIntent);
        finish();
    }

    /**
     * Given an input view, gets the text
     * @param input
     * @return
     */
    private String getPadNameFromInput( TextView input ){
        String padName = (String) input.getText().toString().trim();
        return padName;
    }

    /**
     * Same as previous but with a spinner
     * @param spinner
     * @return
     */
    private String getPadPrefixFromSpinner( Spinner spinner ){
        return server_url_prefixed_list[ spinner.getSelectedItemPosition() ];
    }

    /**
     * Almost the same as the previous one.
     * @param spinner
     * @return
     */
    private String getPadServerFromSpinner( Spinner spinner ){
        return server_url_list[ spinner.getSelectedItemPosition() ];
    }
}

File: app/src/main/java/com/mikifus/padland/NewPadActivity.kt
package com.mikifus.padland

import android.annotation.SuppressLint
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.view.Menu
import android.view.View
import android.webkit.URLUtil
import android.widget.ArrayAdapter
import android.widget.EditText
import android.widget.Spinner
import android.widget.TextView
import android.widget.Toast
import com.mikifus.padland.Models.ServerModel
import com.mikifus.padland.Utils.PadUrl
import java.util.Arrays

/**
 * Allows the user to create a new pad, choosing a name and the host.
 *
 * @author mikifus
 */
open class NewPadActivity : PadLandActivity() {
    private var serverList: Array<String?> = spinnerValueList
    protected var serverUrlList: Array<String?>? = null
    private var serverUrlPrefixedList: Array<String?>? = null
    private var serverModel: ServerModel? = null

    /**
     * onCreate override
     * @param savedInstanceState
     */
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_newpad)
        serverModel = ServerModel(this)
        serverUrlList = serverModel!!.getServerUrlList(this@NewPadActivity)
        serverUrlPrefixedList = serverModel!!.getServerUrlPrefixList(this)
        _setViewEvents()
        _setSpinnerValues()
        _setSpinnerDefaultValue()
    }

    private fun _setViewEvents() {
        val nameInput = findViewById<EditText>(R.id.editText)
        nameInput.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
            override fun afterTextChanged(s: Editable) {
                val name = s.toString()
                if (URLUtil.isValidUrl(name)) {
                    val uri = Uri.parse(name)
                    val padServer = uri.scheme + "://" + uri.host
                    val padName = uri.lastPathSegment
                    for ((c, serverUrl) in serverUrlList!!.withIndex()) {
                        if (serverUrl == padServer) {
                            nameInput.setText(padName)
                            val spinner = findViewById<Spinner>(R.id.spinner)
                            spinner.setSelection(c)
                            Toast.makeText(this@NewPadActivity, getString(R.string.newpad_url_name_success), Toast.LENGTH_LONG).show()
                            return
                        }
                    }
                    Toast.makeText(this@NewPadActivity, getString(R.string.newpad_url_name_warning), Toast.LENGTH_LONG).show()
                }
            }
        })
    }// Load the custom servers

    // Server list to provide a fallback value
//        server_list.getResources().getStringArray( R.array.etherpad_servers_name );
    /**
     * Returns a string with the server names.
     * Includes custom servers.
     * @return String[]
     */
    private val spinnerValueList: Array<String?>
        get() {
            val serverList: Array<String?>
            // Load the custom servers
            val serverModel = ServerModel(this)
            val customServers = serverModel.enabledServerList
            val serverNames = ArrayList<String?>()
            for (server in customServers) {
                serverNames.add(server.name)
            }

            // Server list to provide a fallback value
//        server_list.getResources().getStringArray( R.array.etherpad_servers_name );
            val collection: MutableCollection<String?> = ArrayList()
            collection.addAll(serverNames)
            collection.addAll(listOf(*resources.getStringArray(R.array.etherpad_servers_name)))
            serverList = collection.toTypedArray()
            return serverList
        }

    /**
     * Loads the values on the Spinner which is by deafult empty.
     */
    private fun _setSpinnerValues() {
        val spinner = findViewById<View>(R.id.spinner) as Spinner
        //selected item will look like a spinner set from XML
        val spinnerArrayAdapter = ArrayAdapter(this, android.R.layout.simple_spinner_dropdown_item, serverList)
        spinner.adapter = spinnerArrayAdapter
    }

    /**
     * Gets the default value from user settings and selects it in the
     * spinner. This value can be changed in the settings activity.
     */
    private fun _setSpinnerDefaultValue() {
        val defaultServer = _getDefaultSpinnerValue()

        // We get position and set it as default
        val spinner = findViewById<Spinner>(R.id.spinner)
        val adapter = spinner.adapter as ArrayAdapter<*>
        spinner.setSelection(adapter.getPosition(defaultServer as Nothing?))
    }

    /**
     * Returns a string with the default server name. Being a key in the server
     * array
     * @return
     */
    private fun _getDefaultSpinnerValue(): String? {
        // Getting user preferences
        val context = applicationContext
        val userDetails = context.getSharedPreferences(packageName + "_preferences", MODE_PRIVATE)
        return userDetails.getString("padland_default_server", serverList[0])
    }

    /**
     * Creates the menu
     * @param menu
     * @return
     */
    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        return super.onCreateOptionsMenu(menu, R.menu.new_pad)
    }

    /**
     * Form submit
     */
    @SuppressLint("CutPasteId")
    fun onCreateButtonClick() {
        val padName = getPadNameFromInput(findViewById<View>(R.id.editText) as TextView)
        val padLocalName = getPadNameFromInput(findViewById<View>(R.id.editTextLocalName) as TextView)
        val padPrefix = getPadPrefixFromSpinner(findViewById<View>(R.id.spinner) as Spinner)
        val padServer = getPadServerFromSpinner(findViewById<View>(R.id.spinner) as Spinner)
        Log.d("CREATENEW", padName)
        if (padName.isEmpty()) {
            Toast.makeText(this, getString(R.string.newpad_noname_warning), Toast.LENGTH_LONG).show()
            return
        }
        val padUrl = PadUrl.Builder()
                .padName(padName)
                .padServer(padServer)
                .padPrefix(padPrefix)
                .build()
        if (!URLUtil.isValidUrl(padUrl.string)) {
            Toast.makeText(this, getString(R.string.new_pad_name_invalid), Toast.LENGTH_LONG).show()
            return
        }
        val padViewIntent = Intent(this@NewPadActivity, PadViewActivity::class.java)
        padViewIntent.putExtra("padName", padUrl.padName)
        padViewIntent.putExtra("padLocalName", padLocalName)
        padViewIntent.putExtra("padServer", padUrl.padServer)
        padViewIntent.putExtra("padUrl", padUrl.string)
        startActivity(padViewIntent)
        finish()
    }

    /**
     * Given an input view, gets the text
     * @param input
     * @return
     */
    private fun getPadNameFromInput(input: TextView): String {
        return input.text.toString().trim { it <= ' ' }
    }

    /**
     * Same as previous but with a spinner
     * @param spinner
     * @return
     */
    private fun getPadPrefixFromSpinner(spinner: Spinner): String? {
        return serverUrlPrefixedList!![spinner.selectedItemPosition]
    }

    /**
     * Almost the same as the previous one.
     * @param spinner
     * @return
     */
    private fun getPadServerFromSpinner(spinner: Spinner): String? {
        return serverUrlList!![spinner.selectedItemPosition]
    }
}